File: LeetCode/0018. 4Sum/solution.py
"""
289 / 289 test cases passed.
Runtime: 1056 ms
Memory Usage: 14.9 MB
"""
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        ans = []
        n = len(nums)
        if n < 4:
            return ans
        
        nums.sort()
        for i in range(n - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(i + 1, n - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                l, r = j + 1, n - 1
                while l < r:
                    if nums[i] + nums[j] < target - nums[l] - nums[r]:
                        l += 1
                    elif nums[i] + nums[j] > target - nums[l] - nums[r]:
                        r -= 1
                    else:
                        ans.append([nums[i], nums[j], nums[l], nums[r]])
                        while l < r and nums[l] == nums[l + 1]:
                            l += 1
                        while l < r and nums[r] == nums[r - 1]:
                            r -= 1
                        l += 1
                        r -= 1
        return ans

File: LeetCode/0018. 4Sum/solution.cpp
/**
 * 289 / 289 test cases passed.
 * Runtime: 92 ms
 * Memory Usage: 12.8 MB 
 */
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> ans;
        int n = nums.size();
        if (n < 4) return ans;

        sort(nums.begin(), nums.end());
        for (int i = 0; i <= n - 4; ++ i) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            for (int j = i + 1; j <= n - 3; ++ j) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;
                int l = j + 1, r = n - 1;
                while (l < r) {
                    if (nums[i] + nums[j] < target - nums[l] - nums[r]) {
                        ++ l;
                    } else if (nums[i] + nums[j] > target - nums[l] - nums[r]) {
                        -- r;
                    } else {
                        ans.push_back({nums[i], nums[j], nums[l], nums[r]});
                        while (l < r && nums[l] == nums[l + 1]) ++ l;
                        while (l < r && nums[r] == nums[r - 1]) -- r;
                        ++ l;
                        -- r;
                    } 
                }
            }
        }
        return ans;
    }
};
